crates/jarscanner/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+fn non_anon(file_name: &str) -> bool {
+    !((file_name.contains(r"/$") && file_name.contains(r"$/")) || file_name.contains("$$anon"))
+}
+
+fn not_in_meta(file_name: &str) -> bool {
+    !file_name.starts_with(r"META-INF/")
+}
+
+fn ends_in_class(file_name: &str) -> bool {
+    file_name.ends_with(".class")
+}
+
+fn file_name_to_class_names(file_name: &str) -> Result<Vec<String>, FileNameError> {
+    if non_anon(file_name) && not_in_meta(file_name) && ends_in_class(file_name) {
+        let class_suffix = file_name.strip_suffix(".class").ok_or_else(|| {
+            FileNameError::new(format!("Failed to strip .class suffix for {}", file_name))
+        })?;
+        let final_suffix = class_suffix.strip_suffix("$").unwrap_or(class_suffix);
+
+        let dotted = final_suffix
+            .replace(r"/$", r"/")
+            .replace("$", ".")
+            .replace(r"/", ".");
+        let replace_pkg = dotted.replace(".package", "");
+        if dotted.contains(".package") {
+            Ok(vec![dotted, replace_pkg])
+        } else {
+            Ok(vec![dotted])
+        }
+    } else {
+        Ok(vec![])
+    }
+}
+
+fn read_zip_archive(input_jar: &PathBuf) -> Result<HashSet<String>, Box<dyn Error>> {
+    let file = File::open(input_jar)?;
+    let archive = ZipArchive::new(file)?;
+
+    let mut result = HashSet::new();
+    for file_name in archive.file_names() {
+        match file_name_to_class_names(file_name) {
+            Ok(class_names) => result.extend(class_names.into_iter()),
+            Err(err) => return Err(Box::new(err)),
+        }
+    }
+
+    Ok(result)
+}
+
+fn filter_prefixes(
+    label: &str,
+    classes: HashSet<String>,
+    label_to_allowed_prefixes: &HashMap<String, Vec<String>>,
+) -> HashSet<String> {
+    match label_to_allowed_prefixes.get(label) {
+        Some(prefix_set) => classes
+            .into_iter()
+            .filter(|c| prefix_set.iter().any(|prefix| c.starts_with(prefix)))
+            .collect(),
+        None => classes,
+    }
+}
+
+pub fn process_input(
+    label: &str,
+    input_jar: &PathBuf,
+    relative_path: &str,
+    label_to_allowed_prefixes: &HashMap<String, Vec<String>>,
+) -> Result<ExtractedData, Box<dyn Error>> {
+    let raw_classes = read_zip_archive(input_jar)?;
+    let classes = filter_prefixes(label, raw_classes, &label_to_allowed_prefixes);
+
+    Ok(ExtractedData {
+        label_or_repo_path: label.to_string(),
+        data_blocks: vec![DataBlock {
+            entity_path: relative_path.to_string(),
+            defs: classes,
+            refs: HashSet::new(),
+            bzl_gen_build_commands: HashSet::new(),
+        }],
+    })
+}
+
+pub fn emit_result(
+    target_descriptor: &ExtractedData,
+    output_path: &PathBuf,
+) -> Result<(), Box<dyn Error>> {
+    let json = serde_json::to_string_pretty(target_descriptor)?;
+    let mut file = File::create(output_path)?;
+    file.write_all(json.as_bytes())?;
+    Ok(())
```

Comment:
why do `?;` then `Ok(())`? does write_all not return `()`? Can we add a comment so someone isn't tempted to remove the `?` and trailing `Ok`?

Reply:
Here's what the compiler yells at me if I remove the ? and the `Ok(())`: 
```rust
error[E0308]: mismatched types
   --> jarscanner/src/lib.rs:105:5
    |
102 | ) -> Result<(), Box<dyn Error>> {
    |      -------------------------- expected `Result<(), Box<(dyn std::error::Error + 'static)>>` because of return type
...
105 |     file.write_all(json.as_bytes())
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Result<(), Box<dyn Error>>`, found `Result<(), Error>`
    |
    = note: expected enum `Result<_, Box<(dyn std::error::Error + 'static)>>`
               found enum `Result<_, std::io::Error>`
help: use `?` to coerce and return an appropriate `Err`, and wrap the resulting value in `Ok` so the expression remains of type `Result`
    |
105 |     Ok(file.write_all(json.as_bytes())?)
    |     +++                               ++
```

However, I do think that `Ok(file.write_all(json.as_bytes())?)` is strictly better

Reply:
....or is it? Now second guessing myself. I dunno, i agree with you that it feels like I should just be able to call `write_all` and return the Result from it.

Reply:
I think I'd rather leave what I have than `Ok(file.write_all(json.as_bytes())?)`, having chewed on this for a couple minutes